gnomad_qc/v3/sample_qc/sample_qc.py

```diff
@@ -334,86 +496,351 @@ def apply_stratified_filters(filtering_qc_metrics: List[str]) -> hl.Table:
+    Returns Table keyed by sample with all sample relationships in a set.
+    :param Table relatedness_ht: Table with inferred relationship information output by pc_relate.
+        Keyed by sample pair (i, j).
+    :return: Table keyed by sample (s) with all relationships annotated as a set.
+    :rtype: hl.Table
+    """
+    relatedness_ht = relatedness_ht.filter(relatedness_ht.relationship != UNRELATED)
+    relatedness_ht = relatedness_ht.select("relationship", s=relatedness_ht.i.s).union(
+        relatedness_ht.select("relationship", s=relatedness_ht.j.s)
+    )
+    relatedness_ht = relatedness_ht.group_by(relatedness_ht.s).aggregate(
+        relationships=hl.agg.collect_as_set(relatedness_ht.relationship)
+    )
+    return relatedness_ht
+
+
+def get_relationship_filter_expr(
+    hard_filtered_expr: hl.expr.BooleanExpression,
+    relationship: str,
+    relationship_set: hl.expr.SetExpression,
+) -> hl.expr.builders.CaseBuilder:
+    """
+    Returns case statement to populate relatedness filters in sample_filters struct
+    :param hl.expr.BooleanExpression hard_filtered_expr: Boolean for whether sample was hard filtered.
+    :param str relationship: Relationship to check for. One of DUPLICATE_OR_TWINS, PARENT_CHILD, or SIBLINGS.
+    :param hl.expr.SetExpression relationship_set: Set containing all possible relationship strings for sample.
+    :return: Case statement used to population sample_filters related filter field.
+    :rtype: hl.expr.builders.CaseBuilder
+    """
+    return (
+        hl.case()
+        .when(hard_filtered_expr, hl.null(hl.tbool))
+        .when(hl.is_defined(relationship_set), relationship_set.contains(relationship))
+        .default(False)
+    )
+
+
+def compare_row_counts(ht1: hl.Table, ht2: hl.Table) -> bool:
+    """
+    Checks if row counts in two Tables are the same
+    :param Table ht1: First Table to be checked
+    :param Table ht2: Second Table to be checked
+    :return: Whether the row counts are the same
+    :rtype: bool
+    """
+    r_count1 = ht1.count()
+    r_count2 = ht2.count()
+    logger.info(f"{r_count1} rows in left table; {r_count2} rows in right table")
+    return r_count1 == r_count2
+
+
+def join_tables(
+        left_ht: hl.Table, left_key: str, right_ht: hl.Table, right_key: str, join_type: str
+) -> hl.Table:
+    """
+    Joins left and right tables using specified keys and join types and returns result.
+    Also prints warning if sample counts are not the same.
+    :param Table left_ht: Left Table to be joined
+    :param str left_key: Key of left Table
+    :param Table right_ht: Right Table to be joined
+    :param str right_key: Key of right Table
+    :param str join_type: Type of join
+    :return: Table with annotations
+    :rtype: Table
+    """
+    sample_count_match = compare_row_counts(left_ht, right_ht)
+    if not sample_count_match:
+        logger.warning("Sample counts in left and right tables do not match!")
```

Comment:
If the two table row counts don't match, do we want to proceed with the join? If we write out the samples that don't match, then could we prompt the user to proceed?

Reply:
We do not always expect them to match. We added these to the UKBB join just to look over as a sanity check where tables didn't match, but I can change to only warn for the ones that we don't necessarily expect to. I don't think I want to add in a prompt because I don't want the cluster to keep running, but if I only check for the ones I expect to be the same then I could have it fail instead of just warning. I'm happy either way, but kept as a warning for now